File: apps/desktop/src/services/client/resources/thread.ts
import type { ThreadsResponse } from "../../../types";
import { call } from "../invoke";
import type { EditCommentBody, PostCommentBody } from "../types";

export class ThreadResource {
  get(ws: string, taskId: string) {
    return call<ThreadsResponse>("thread_get", { workspaceId: ws, taskId });
  }

  postComment(ws: string, taskId: string, body: PostCommentBody) {
    return call<void>("thread_append", {
      workspaceId: ws,
      taskId,
      content: body.content,
      variant: body.variant,
    });
  }

  editComment(ws: string, taskId: string, entryId: number, body: EditCommentBody) {
    return call<void>("thread_edit", {
      workspaceId: ws,
      taskId,
      entryId,
      content: body.content,
      author: body.author,
    });
  }

  deleteComment(ws: string, taskId: string, entryId: number, author: string) {
    return call<void>("thread_delete", { workspaceId: ws, taskId, entryId, author });
  }
}

File: apps/desktop/src-tauri/src/grpc/thread.rs
use modula_client::{AppendEntry, ModulaClient};
use modula_types::ThreadBundle;
use tauri::State;

#[tauri::command]
pub async fn thread_get(
    engine: State<'_, ModulaClient>,
    workspace_id: String,
    task_id: String,
) -> Result<ThreadBundle, String> {
    Ok(engine.get_threads(&workspace_id, &task_id).await?)
}

/// Post a human comment. The desktop only ever appends `human`/`comment`
/// entries; agent verdicts and rework summaries come from the CLI.
#[tauri::command]
pub async fn thread_append(
    engine: State<'_, ModulaClient>,
    workspace_id: String,
    task_id: String,
    content: String,
    variant: Option<String>,
) -> Result<(), String> {
    engine
        .append_entry(AppendEntry {
            workspace_id,
            task_id,
            content,
            author: "human".into(),
            kind: "comment".into(),
            variant_id: variant,
            round: None,
            verdict: None,
            affected_variants: Vec::new(),
        })
        .await?;
    Ok(())
}

#[tauri::command]
pub async fn thread_edit(
    engine: State<'_, ModulaClient>,
    workspace_id: String,
    task_id: String,
    entry_id: i64,
    content: String,
    author: String,
) -> Result<(), String> {
    engine
        .edit_entry(&workspace_id, &task_id, entry_id, &content, &author)
        .await?;
    Ok(())
}

#[tauri::command]
pub async fn thread_delete(
    engine: State<'_, ModulaClient>,
    workspace_id: String,
    task_id: String,
    entry_id: i64,
    author: String,
) -> Result<(), String> {
    engine
        .delete_entry(&workspace_id, &task_id, entry_id, &author)
        .await?;
    Ok(())
}
